pkg/reconciler/event_test.go

```diff
@@ -127,43 +122,70 @@ func TestEmitEvent(t *testing.T) {
 			Status:  corev1.ConditionFalse,
 			Message: "really bad",
 		},
-		expectEvent:   true,
-		expectedEvent: "Warning Failed really bad",
+		wantEvent: "Warning Failed really bad",
 	}, {
 		name:   "nil to false",
 		before: nil,
 		after: &apis.Condition{
 			Type:   apis.ConditionSucceeded,
 			Status: corev1.ConditionFalse,
 		},
-		expectEvent:   true,
-		expectedEvent: "Warning Failed ",
+		wantEvent: "Warning Failed ",
 	}}
 
 	for _, ts := range testcases {
 		fr := record.NewFakeRecorder(1)
 		tr := &corev1.Pod{}
 		EmitEvent(fr, ts.before, ts.after, tr)
-		timer := time.NewTimer(1 * time.Second)
 
-		select {
-		case event := <-fr.Events:
-			if event == "" {
-				// The fake recorder reported empty, it should not happen
-				t.Fatalf("Expected event but got empty for %s", ts.name)
-			}
-			if !ts.expectEvent {
-				// The fake recorder reported an event which we did not expect
-				t.Errorf("Unxpected event \"%s\" but got one for %s", event, ts.name)
-			}
-			if !(event == ts.expectedEvent) {
-				t.Errorf("Expected event \"%s\" but got \"%s\" instead for %s", ts.expectedEvent, event, ts.name)
-			}
-		case <-timer.C:
-			if ts.expectEvent {
-				// The fake recorder did not report, the timer timeout expired
-				t.Errorf("Expected event but got none for %s", ts.name)
-			}
+		err := checkEvents(fr, ts.name, ts.wantEvent)
+		if err != nil {
+			t.Errorf(err.Error())
 		}
 	}
 }
+
+func TestEmitErrorEvent(t *testing.T) {
+	testcases := []struct {
+		name      string
+		err       error
+		wantEvent string
+	}{{
+		name:      "with error",
+		err:       errors.New("something went wrong"),
+		wantEvent: "Warning Error something went wrong",
+	}, {
+		name:      "without error",
+		err:       nil,
+		wantEvent: "",
+	}}
+
+	for _, ts := range testcases {
+		fr := record.NewFakeRecorder(1)
+		tr := &corev1.Pod{}
+		EmitErrorEvent(fr, ts.err, tr)
+
+		err := checkEvents(fr, ts.name, ts.wantEvent)
+		if err != nil {
+			t.Errorf(err.Error())
+		}
+	}
+}
+
+func checkEvents(fr *record.FakeRecorder, testName string, wantEvent string) error {
+	timer := time.NewTimer(1 * time.Second)
+	select {
+	case event := <-fr.Events:
+		if wantEvent == "" {
+			return fmt.Errorf("received event \"%s\" for %s but none expected", event, testName)
+		}
+		if !(strings.HasPrefix(event, wantEvent)) {
+			return fmt.Errorf("expected event \"%s\" but got \"%s\" instead for %s", wantEvent, event, testName)
+		}
+	case <-timer.C:
+		if wantEvent != "" {
+			return fmt.Errorf("received no events for %s but %s expected", testName, wantEvent)
+		}
+	}
+	return nil
+}
```

Comment:
when this fails, what line number does the test failure produce? would it be useful to use `t.Helper` here?

Reply:
Yes, it would be nice, I'll add `t` to the method interface then